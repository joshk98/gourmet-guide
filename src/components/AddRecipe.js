import React, { useState } from "react";
import axios from "axios";
import "../styles/addrecipe.css";

const AddRecipe = () => {
  const initialState = {
    fields: {
      title: "",
      cuisine: "",
      dietaryRequirements: [],
      ingredients: [{ name: "", measurement: "" }],
      instructions: "",
      prepTime: "",
      cookingTime: "",
      servings: "",
    },
    alert: {
      message: "",
      isSuccess: false,
    },
  };

  const [fields, setFields] = useState(initialState.fields);
  const [alert, setAlert] = useState(initialState.alert);

  const handleAddRecipe = (event) => {
    event.preventDefault();
    if (
      fields.title.trim() === "" ||
      fields.ingredients.length === 0 ||
      fields.instructions.trim() === "" ||
      fields.prepTime <= 0 ||
      fields.cookingTime <= 0 ||
      fields.servings <= 0
    ) {
      setAlert({
        message: "Please fill in all required fields.",
        isSuccess: false,
      });
      return;
    }

    setAlert({ message: "", isSuccess: false });

    axios
      .post("http://localhost:3000/api/v1/Recipe", fields)
      .then(() =>
        setAlert({
          message: "Recipe Added",
          isSuccess: true,
        })
      )
      .catch(() =>
        setAlert({
          message: "Server error. Please try again later.",
          isSuccess: false,
        })
      );
  };

  const handleFieldChange = (event) => {
    const { name, value } = event.target;

    if (name === "dietaryRequirements") {
      const selectedOptions = Array.from(
        event.target.selectedOptions,
        (option) => option.value
      );
      setFields({ ...fields, [name]: selectedOptions });
    } else {
      setFields({ ...fields, [name]: value });
    }
  };

  const handleAddIngredient = () => {
    setFields({
      ...fields,
      ingredients: [...fields.ingredients, { name: "", measurement: "" }],
    });
  };

  const handleRemoveIngredient = (index) => {
    const updatedIngredients = [...fields.ingredients];
    updatedIngredients.splice(index, 1);
    setFields({ ...fields, ingredients: updatedIngredients });
  };

  const handleIngredientChange = (index, field, value) => {
    const updatedIngredients = [...fields.ingredients];
    updatedIngredients[index][field] = value;
    setFields({ ...fields, ingredients: updatedIngredients });
  };

  return (
    <div className="add-recipe">
      <form onSubmit={handleAddRecipe}>
        <div>
          <label htmlFor="title">
            Title:
            <input
              type="text"
              id="title"
              name="title"
              value={fields.title}
              onChange={handleFieldChange}
              placeholder="Enter the title"
            />
          </label>
        </div>
        <div>
          <label htmlFor="cuisine">
            Cuisine:
            <select
              id="cuisine"
              name="cuisine"
              value={fields.cuisine}
              onChange={handleFieldChange}
            >
              <option value="" disabled>
                --Select a cuisine--
              </option>
              <option value="Italian">Italian</option>
              <option value="Chinese">Chinese</option>
              <option value="Japanese">Japanese</option>
              <option value="Indian">Indian</option>
              <option value="Mexican">Mexican</option>
              <option value="French">French</option>
              <option value="Thai">Thai</option>
              <option value="Spanish">Spanish</option>
              <option value="Middle Eastern">Middle Eastern</option>
              <option value="Korean">Korean</option>
            </select>
          </label>
        </div>
        <div>
          <label htmlFor="dietaryRequirements">
            Dietary Requirements:
            <select
              id="dietaryRequirements"
              name="dietaryRequirements"
              value={fields.dietaryRequirements}
              onChange={handleFieldChange}
            >
              <option value="" disabled>
                --Select a dietary requirement--
              </option>
              <option value="Vegetarian">Vegetarian</option>
              <option value="Vegan">Vegan</option>
              <option value="Gluten-Free">Gluten-Free</option>
              <option value="Dairy-Free">Dairy-Free</option>
              <option value="Pescatarian">Pescatarian</option>
            </select>
          </label>
        </div>
        <div>
          <label htmlFor="ingredients">
            Ingredients:
            {fields.ingredients.map((ingredient, index) => (
              <div
                key={`${ingredient.name}-${ingredient.measurement}`}
                className="ingredient-field"
              >
                <input
                  type="text"
                  placeholder="Ingredient Name"
                  value={ingredient.name}
                  onChange={(e) =>
                    handleIngredientChange(index, "name", e.target.value)
                  }
                />
                <input
                  type="text"
                  placeholder="Measurement Unit"
                  value={ingredient.measurement}
                  onChange={(e) =>
                    handleIngredientChange(index, "measurement", e.target.value)
                  }
                />
                {index !== 0 && (
                  <button
                    type="button"
                    onClick={() => handleRemoveIngredient(index)}
                  >
                    Remove
                  </button>
                )}
              </div>
            ))}
            <button type="button" onClick={handleAddIngredient}>
              Add Ingredient
            </button>
          </label>
        </div>
        <div>
          <label htmlFor="instructions">
            Instructions:
            <textarea
              id="instructions"
              name="instructions"
              value={fields.instructions}
              onChange={handleFieldChange}
              placeholder="Enter the instructions"
            />
          </label>
        </div>
        <div>
          <label htmlFor="prepTime">
            Prep Time (minutes):
            <input
              type="number"
              id="prepTime"
              name="prepTime"
              value={fields.prepTime}
              onChange={handleFieldChange}
              placeholder="Enter the prep time"
            />
          </label>
        </div>
        <div>
          <label htmlFor="cookingTime">
            Cooking Time (minutes):
            <input
              type="number"
              id="cookingTime"
              name="cookingTime"
              value={fields.cookingTime}
              onChange={handleFieldChange}
              placeholder="Enter the cooking time"
            />
          </label>
        </div>
        <div>
          <label htmlFor="servings">
            Servings:
            <input
              type="number"
              id="servings"
              name="servings"
              value={fields.servings}
              onChange={handleFieldChange}
              placeholder="Enter the number of servings"
            />
          </label>
        </div>
        <div>
          <button type="submit">Add</button>
        </div>
      </form>
    </div>
  );
};

export default AddRecipe;
