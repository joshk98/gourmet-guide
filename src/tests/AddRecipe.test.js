import React from "react";
import { render, fireEvent } from "@testing-library/react";
import AddRecipe from "../components/AddRecipe";

describe("AddRecipe", () => {
  it("renders correctly", () => {
    const { asFragment } = render(<AddRecipe />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("updates form input correctly - Title", () => {
    const { getByLabelText } = render(<AddRecipe />);
    const titleInput = getByLabelText("Title:");
    fireEvent.change(titleInput, { target: { value: "Pizza Margherita" } });

    expect(titleInput.value).toBe("Pizza Margherita");
  });

  it("updates form input correctly - Cuisine", () => {
    const { getByLabelText } = render(<AddRecipe />);
    const cuisineSelect = getByLabelText("Cuisine:");
    fireEvent.change(cuisineSelect, { target: { value: "Italian" } });

    expect(cuisineSelect.value).toBe("Italian");
  });

  it("updates form input correctly - Dietary Requirements", () => {
    const { getByLabelText } = render(<AddRecipe />);
    const dietaryRequirementsSelect = getByLabelText("Dietary Requirements:");
    fireEvent.change(dietaryRequirementsSelect, { target: { value: "None" } });

    expect(dietaryRequirementsSelect.value).toBe("None");
  });

  it("updates form input correctly - Instructions", () => {
    const { getByLabelText } = render(<AddRecipe />);
    const instructionsTextarea = getByLabelText("Instructions:");
    fireEvent.change(instructionsTextarea, {
      target: {
        value:
          "1. Preheat the oven to 220°C.\n2. Roll out the pizza dough.\n3. Spread tomato sauce on the dough.\n4. Add slices of mozzarella cheese and fresh basil leaves.\n5. Bake in the oven for 12-15 minutes or until the crust is golden and the cheese is bubbly and slightly browned.",
      },
    });

    expect(instructionsTextarea.value).toBe(
      "1. Preheat the oven to 220°C.\n2. Roll out the pizza dough.\n3. Spread tomato sauce on the dough.\n4. Add slices of mozzarella cheese and fresh basil leaves.\n5. Bake in the oven for 12-15 minutes or until the crust is golden and the cheese is bubbly and slightly browned."
    );
  });

  it("updates form input correctly - Prep Time", () => {
    const { getByLabelText } = render(<AddRecipe />);
    const prepTimeInput = getByLabelText("Prep Time (minutes):");
    fireEvent.change(prepTimeInput, { target: { value: "20" } });

    expect(prepTimeInput.value).toBe("20");
  });

  it("updates form input correctly - Cooking Time", () => {
    const { getByLabelText } = render(<AddRecipe />);
    const cookingTimeInput = getByLabelText("Cooking Time (minutes):");
    fireEvent.change(cookingTimeInput, { target: { value: "15" } });

    expect(cookingTimeInput.value).toBe("15");
  });

  it("updates form input correctly - Servings", () => {
    const { getByLabelText } = render(<AddRecipe />);
    const servingsInput = getByLabelText("Servings:");
    fireEvent.change(servingsInput, { target: { value: "2" } });

    expect(servingsInput.value).toBe("2");
  });
});

it("adds new ingredient field on button click", () => {
  const { getByText, getAllByPlaceholderText } = render(<AddRecipe />);

  const initialIngredientFields = getAllByPlaceholderText("Ingredient Name");
  const initialMeasurementFields = getAllByPlaceholderText("Measurement Unit");
  expect(initialIngredientFields.length).toBe(1);
  expect(initialMeasurementFields.length).toBe(1);

  const addIngredientButton = getByText("Add Ingredient");
  fireEvent.click(addIngredientButton);

  const updatedIngredientFields = getAllByPlaceholderText("Ingredient Name");
  const updatedMeasurementFields = getAllByPlaceholderText("Measurement Unit");
  expect(updatedIngredientFields.length).toBe(2);
  expect(updatedMeasurementFields.length).toBe(2);
});

it("removes ingredient field on button click", () => {
  const { getByText, getByPlaceholderText, queryByText } = render(
    <AddRecipe />
  );

  const initialIngredientField = getByPlaceholderText("Ingredient Name");
  const initialMeasurementField = getByPlaceholderText("Measurement Unit");
  expect(initialIngredientField).toBeInTheDocument();
  expect(initialMeasurementField).toBeInTheDocument();

  const addIngredientButton = getByText("Add Ingredient");
  fireEvent.click(addIngredientButton);

  const removeButton = getByText("Remove");
  expect(removeButton).toBeInTheDocument();

  fireEvent.click(removeButton);

  const updatedIngredientField = queryByText("Ingredient Name");
  const updatedMeasurementField = queryByText("Measurement Unit");
  expect(updatedIngredientField).toBeNull();
  expect(updatedMeasurementField).toBeNull();
});

// it("submits form data correctly", () => {
//   const { getByLabelText, getByText } = render(<AddRecipe />);

//   // Fill out the form
//   const titleInput = getByLabelText("Title:");
//   const cuisineInput = getByLabelText("Cuisine:");
//   const dietaryRequirementsInput = getByLabelText("Dietary Requirements:");
//   const instructionsInput = getByLabelText("Instructions:");
//   const prepTimeInput = getByLabelText("Prep Time (minutes):");
//   const cookingTimeInput = getByLabelText("Cooking Time (minutes):");
//   const servingsInput = getByLabelText("Servings:");

//   fireEvent.change(titleInput, { target: { value: "Delicious Pasta" } });
//   fireEvent.change(cuisineInput, { target: { value: "Italian" } });
//   fireEvent.change(dietaryRequirementsInput, {
//     target: { value: "Vegetarian" },
//   });
//   fireEvent.change(instructionsInput, {
//     target: { value: "Cook pasta and add sauce" },
//   });
//   fireEvent.change(prepTimeInput, { target: { value: "10" } });
//   fireEvent.change(cookingTimeInput, { target: { value: "20" } });
//   fireEvent.change(servingsInput, { target: { value: "2" } });

//   // Add an ingredient
//   const addIngredientButton = getByText("Add Ingredient");
//   fireEvent.click(addIngredientButton);

//   const ingredientNameInput = getByLabelText("Ingredient Name");
//   const measurementInput = getByLabelText("Measurement Unit");
//   fireEvent.change(ingredientNameInput, { target: { value: "Pasta" } });
//   fireEvent.change(measurementInput, { target: { value: "500g" } });

//   // Submit the form
//   const submitButton = getByText("Add");
//   fireEvent.click(submitButton);

//   // Verify the form data is submitted correctly
//   expect(submitButton).not.toBeInTheDocument(); // Form should disappear after submission
//   // You can add more assertions here to check if the submitted data is correct, e.g. by checking if it appears on the page or calling a function that verifies the data was sent to the server.
//   // For this example, let's assume the form data is displayed on the page after submission:
//   const submittedData = getByText("Delicious Pasta");
//   expect(submittedData).toBeInTheDocument();

//   // Check if the form fields are cleared after submission
//   expect(titleInput.value).toBe("");
//   expect(cuisineInput.value).toBe("");
//   expect(dietaryRequirementsInput.value).toBe("");
//   expect(instructionsInput.value).toBe("");
//   expect(prepTimeInput.value).toBe("");
//   expect(cookingTimeInput.value).toBe("");
//   expect(servingsInput.value).toBe("");
//   expect(ingredientNameInput.value).toBe("");
//   expect(measurementInput.value).toBe("");
// });
