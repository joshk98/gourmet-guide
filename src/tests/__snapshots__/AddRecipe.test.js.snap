// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRecipe renders correctly 1`] = `
<DocumentFragment>
  <div
    class="add-recipe"
  >
    <form>
      <div>
        <label
          for="title"
        >
          Title:
          <input
            id="title"
            name="title"
            placeholder="Enter the title"
            type="text"
            value=""
          />
        </label>
      </div>
      <div>
        <label
          for="cuisine"
        >
          Cuisine:
          <select
            id="cuisine"
            name="cuisine"
          >
            <option
              disabled=""
              value=""
            >
              --Select a cuisine--
            </option>
            <option
              value="Italian"
            >
              Italian
            </option>
            <option
              value="Chinese"
            >
              Chinese
            </option>
            <option
              value="Japanese"
            >
              Japanese
            </option>
            <option
              value="Indian"
            >
              Indian
            </option>
            <option
              value="Mexican"
            >
              Mexican
            </option>
            <option
              value="French"
            >
              French
            </option>
            <option
              value="Thai"
            >
              Thai
            </option>
            <option
              value="Spanish"
            >
              Spanish
            </option>
            <option
              value="Middle Eastern"
            >
              Middle Eastern
            </option>
            <option
              value="Korean"
            >
              Korean
            </option>
          </select>
        </label>
      </div>
      <div>
        <label
          for="dietaryRequirements"
        >
          Dietary Requirements:
          <select
            id="dietaryRequirements"
            name="dietaryRequirements"
          >
            <option
              disabled=""
              value=""
            >
              --Select a dietary requirement--
            </option>
            <option
              value="None"
            >
              None
            </option>
            <option
              value="Vegetarian"
            >
              Vegetarian
            </option>
            <option
              value="Vegan"
            >
              Vegan
            </option>
            <option
              value="Gluten-Free"
            >
              Gluten-Free
            </option>
            <option
              value="Dairy-Free"
            >
              Dairy-Free
            </option>
            <option
              value="Pescatarian"
            >
              Pescatarian
            </option>
          </select>
        </label>
      </div>
      <div>
        <label
          for="ingredients"
        >
          Ingredients:
          <div
            class="ingredient-field"
          >
            <input
              id="ingredientName-0"
              placeholder="Ingredient Name"
              type="text"
              value=""
            />
            <input
              id="measurement-0"
              placeholder="Measurement Unit"
              type="text"
              value=""
            />
          </div>
          <button
            type="button"
          >
            Add Ingredient
          </button>
        </label>
      </div>
      <div>
        <label
          for="instructions"
        >
          Instructions:
          <textarea
            id="instructions"
            name="instructions"
            placeholder="Enter the instructions"
          />
        </label>
      </div>
      <div>
        <label
          for="prepTime"
        >
          Prep Time (minutes):
          <input
            id="prepTime"
            name="prepTime"
            placeholder="Enter the prep time"
            type="number"
            value=""
          />
        </label>
      </div>
      <div>
        <label
          for="cookingTime"
        >
          Cooking Time (minutes):
          <input
            id="cookingTime"
            name="cookingTime"
            placeholder="Enter the cooking time"
            type="number"
            value=""
          />
        </label>
      </div>
      <div>
        <label
          for="servings"
        >
          Servings:
          <input
            id="servings"
            name="servings"
            placeholder="Enter the number of servings"
            type="number"
            value=""
          />
        </label>
      </div>
      <div>
        <button
          type="submit"
        >
          Add
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
